### PR TITLE
add support for bracketed IPv6 addresses in the IRI authority component

### PR DESCRIPTION
**Pull request by gibson042**
Ref #62

includes changes from #63

`benchstat` vs. master@68fd825f4d226a5b043f697ed9c2c29011c4c9ac

    name                         old time/op    new time/op    delta
    Strict_none-4                  45.7µs ±27%    41.8µs ± 4%     ~     (p=0.315 n=10+8)
    Strict_many-4                  36.0µs ± 4%    36.8µs ± 6%     ~     (p=0.156 n=9+10)
    Relaxed_none-4                 60.8µs ±19%    68.2µs ±17%  +12.19%  (p=0.011 n=10+10)
    Relaxed_many-4                 50.1µs ± 7%    56.8µs ±18%  +13.37%  (p=0.000 n=9+10)
    StrictMatchingScheme_none-4     775ns ±39%     743ns ±24%     ~     (p=0.739 n=10+10)
    StrictMatchingScheme_many-4    1.77µs ± 9%    1.95µs ±17%     ~     (p=0.053 n=9+10)

    name                         old speed      new speed      delta
    Strict_none-4                1.73MB/s ±22%  1.87MB/s ± 4%     ~     (p=0.303 n=10+8)
    Strict_many-4                2.69MB/s ± 4%  2.64MB/s ± 6%     ~     (p=0.138 n=9+10)
    Relaxed_none-4               1.29MB/s ±16%  1.16MB/s ±15%  -10.60%  (p=0.011 n=10+10)
    Relaxed_many-4               1.94MB/s ± 6%  1.72MB/s ±16%  -11.34%  (p=0.000 n=9+10)
    StrictMatchingScheme_none-4   103MB/s ±30%   106MB/s ±20%     ~     (p=0.739 n=10+10)
    StrictMatchingScheme_many-4  53.9MB/s ±19%  50.4MB/s ±18%     ~     (p=0.123 n=10+10)

    name                         old alloc/op   new alloc/op   delta
    Strict_none-4                   0.00B          0.00B          ~     (all equal)
    Strict_many-4                    230B ± 7%      233B ± 4%     ~     (p=0.790 n=10+10)
    Relaxed_none-4                 8.80B ±241%     0.00B          ~     (p=0.173 n=10+9)
    Relaxed_many-4                   240B ± 0%      240B ± 0%     ~     (all equal)
    StrictMatchingScheme_none-4     0.00B          0.00B          ~     (all equal)
    StrictMatchingScheme_many-4      192B ± 0%      192B ± 0%     ~     (all equal)

    name                         old allocs/op  new allocs/op  delta
    Strict_none-4                    0.00           0.00          ~     (all equal)
    Strict_many-4                    5.00 ± 0%      5.00 ± 0%     ~     (all equal)
    Relaxed_none-4                   0.00           0.00          ~     (all equal)
    Relaxed_many-4                   6.00 ± 0%      6.00 ± 0%     ~     (all equal)
    StrictMatchingScheme_none-4      0.00           0.00          ~     (all equal)
    StrictMatchingScheme_many-4      3.00 ± 0%      3.00 ± 0%     ~     (all equal)

### CODE REVIEW

File: xurls_test.go [hunk 1]
```diff
@@ -272,6 +350,10 @@ func TestRegexes(t *testing.T) {
 		{`3ffe:2a00:100:7031::1`, nil},
 		{`test.foo.com:8080/path`, nil},
 		{`foo@bar.com`, nil},
+
+		// An IP address in URI host position must be bracketed unless it is IPv4.
+		// https://www.rfc-editor.org/rfc/rfc3986#section-3.2.2
+		//{`http://1080::8:800:200c:417a/path`, nil},
```
Comment (mvdan) [inline xurls_test.go:356]: should this be commented out?

File: xurls.go [hunk 2]
```diff
@@ -34,17 +36,46 @@ const (
...
+		// up to 5 final chomps.
+		`(?:` + h4 + `:){2}(?:(?::` + h4 + `){0,3}:` + ipv4Addr + `|(?::` + h4 + `){1,5}|:)|` +
+		// 1 chomp, followed by an elision with optional IPv4 (preceded by up to 4 chomps) or
+		// up to 6 final chomps.
+		`(?:` + h4 + `:){1}(?:(?::` + h4 + `){0,4}:` + ipv4Addr + `|(?::` + h4 + `){1,6}|:)|` +
+		// elision, followed by optional IPv4 (preceded by up to 5 chomps) or
+		// up to 7 final chomps.
+		// `:` is an intentionally omitted alternative, to avoid matching `::`.
+		`:(?:(?::` + h4 + `){0,5}:` + ipv4Addr + `|(?::` + h4 + `){1,7})` +
+		`)`
```
Comment (mvdan) [inline xurls.go:75]: nice; I had thought about a similar refactor when reviewing a previous PR.

**Review by mvdan — APPROVED** [review]
A 10% slow-down on relaxed to not match some invalid URLs seems fine. Thanks again!

File: xurls_test.go (hunk 1 above)
Comment (gibson042) [reply xurls_test.go:356]: I'd love to uncomment it, but the current code still incorrectly accepts bare IPv6 addresses as IRI hostname components, and I haven't yet found a good way to fix that.

Comment (mvdan) [reply xurls_test.go:356]: fair; perhaps add a TODO to clarify

Comment (gibson042) [reply xurls_test.go:358]: Of course; done.

**Review by mvdan — APPROVED** [review]
Great!

**Pull request merged**